File: backend/prompts/taxonomy.py
"""Canonical labels for communication signals and logical fallacies.

Single source of truth used by the prompt builder, the schema normaliser, tests, the eval
scripts and the fine-tune dataset generator. Definitions and detection cues live in
docs/TECHNIQUES.md; keep the two in sync.
"""

from __future__ import annotations

TACTICS: tuple[str, ...] = (
    "Outrage Farming",
    "Scapegoating",
    "Fear-mongering",
    "Dog Whistle",
    "Us-vs-Them Framing",
    "Dehumanization",
    "Emotional Bait",
    "Manufactured Urgency",
    "Cherry Picking",
    "Whataboutism",
    "Gish Gallop",
    "Astroturfing",
)

FALLACIES: tuple[str, ...] = (
    "False Dilemma",
    "Ad Hominem",
    "Hasty Generalization",
    "Slippery Slope",
    "Appeal to Fear",
    "Appeal to Emotion",
    "Straw Man",
    "Middle Ground",
    "Bandwagon",
    "Appeal to Authority",
    "Loaded Question",
    "No True Scotsman",
    "Post Hoc",
    "Motte and Bailey",
)

# Labels with reputational risk if wrong: the UI hides them below this confidence.
HIGH_RISK_TACTICS: frozenset[str] = frozenset({"Dog Whistle", "Scapegoating", "Dehumanization"})
HIGH_RISK_MIN_CONFIDENCE = 0.6

# Accepted spellings/synonyms → canonical. Lower-cased keys.
_SYNONYMS: dict[str, str] = {
    # tactics
    "outrage bait": "Outrage Farming",
    "rage bait": "Outrage Farming",
    "ragebait": "Outrage Farming",
    "fear mongering": "Fear-mongering",
    "fearmongering": "Fear-mongering",
    "fear appeal": "Fear-mongering",
    "dog-whistle": "Dog Whistle",
    "dog whistling": "Dog Whistle",
    "dogwhistle": "Dog Whistle",
    "in-group/out-group": "Us-vs-Them Framing",
    "in-group vs out-group": "Us-vs-Them Framing",
    "us vs them": "Us-vs-Them Framing",
    "us-vs-them": "Us-vs-Them Framing",
    "othering": "Us-vs-Them Framing",
    "tribalism": "Us-vs-Them Framing",
    "dehumanisation": "Dehumanization",
    "engagement bait": "Emotional Bait",
    "urgency": "Manufactured Urgency",
    "false urgency": "Manufactured Urgency",
    "cherry-picking": "Cherry Picking",
    "selective evidence": "Cherry Picking",
    "whataboutery": "Whataboutism",
    "tu quoque": "Ad Hominem",
    "manufactured consensus": "Astroturfing",
    # fallacies
    "false dichotomy": "False Dilemma",
    "either/or": "False Dilemma",
    "black-and-white thinking": "False Dilemma",
    "personal attack": "Ad Hominem",
    "overgeneralization": "Hasty Generalization",
    "over-generalization": "Hasty Generalization",
    "sweeping generalization": "Hasty Generalization",
    "appeal to popularity": "Bandwagon",
    "ad populum": "Bandwagon",
    "argumentum ad populum": "Bandwagon",
    "golden mean": "Middle Ground",
    "argument to moderation": "Middle Ground",
    "false cause": "Post Hoc",
    "post hoc ergo propter hoc": "Post Hoc",
    "strawman": "Straw Man",
    "straw-man": "Straw Man",
    "misplaced authority": "Appeal to Authority",
    "complex question": "Loaded Question",
    "appeal to pity": "Appeal to Emotion",
}

_CANONICAL_LOWER: dict[str, str] = {t.lower(): t for t in TACTICS + FALLACIES}


def normalize_label(name: str) -> str:
    """Map a model-produced label to its canonical spelling.

    Unknown labels are kept but prefixed with "Other: " so they are visible in eval output
    and can be promoted to the canonical list later.
    """
    key = name.strip().strip('"').lower()
    if key.startswith("other:"):
        key = key[6:].strip()
    if key in _CANONICAL_LOWER:
        return _CANONICAL_LOWER[key]
    if key in _SYNONYMS:
        return _SYNONYMS[key]
    return f"Other: {name.strip()}"


def is_canonical(name: str) -> bool:
    return name in TACTICS or name in FALLACIES


def prompt_block() -> str:
    """The allowed-vocabulary block inserted into the system prompt."""
    return (
        "ALLOWED communication_signals names: " + "; ".join(TACTICS) + ".\n"
        "ALLOWED logical_fallacies names: " + "; ".join(FALLACIES) + ".\n"
        'Use these spellings exactly. If a pattern is not listed, use "Other: <short name>".'
    )

File: extension/content/taxonomy.js
// Display rules for the claim verdict and the rhetorical signal labels.
// Mirrors backend/prompts/taxonomy.py and the Verdict literal in schemas/analysis_schema.py.
//
// The backend normalises every label before it reaches us, so this file only decides how
// something looks, never whether it is valid.
//
// v3 merged manipulation tactics and logical fallacies into one vocabulary and dropped the
// per-signal confidence score, so there is no confidence threshold to filter on any more.

(() => {
  // Wording and colour for each verdict. The verdict is about the extracted claim only -
  // never about the post - so no label here says "true", "false" or "fake".
  const VERDICTS = {
    supported:           { label: "Supported",           tone: "green" },
    partially_supported: { label: "Partially supported", tone: "amber" },
    unsupported:         { label: "Unsupported",         tone: "red"   },
    unverifiable:        { label: "Unverifiable",        tone: "grey"  },
    no_factual_claim:    { label: "No factual claim",    tone: "grey"  },
  };

  const FALLBACK = { label: "Unverifiable", tone: "grey" };

  // Reputational risk if the model is wrong about these. backend/prompts/taxonomy.py asks
  // the client to render them cautiously; we mark them rather than hide them, since v3
  // gives us no confidence value to threshold on.
  const HIGH_RISK = new Set(["Dog Whistle", "Scapegoating", "Dehumanization"]);

  // What each signal name means, in one line, for the card's hover definition. Mirrors CUES in
  // backend/prompts/taxonomy.py and scripts/check_extension.py compares the two: a badge on its
  // own is an accusation in jargon, and a reader who does not know what "Dog Whistle" means
  // cannot agree or disagree with it.
  const MEANINGS = {
    "Loaded Language": "An emotionally charged word chosen over a neutral one.",
    "Outrage Farming": "Framed to provoke sharing and anger rather than to inform.",
    "Scapegoating": "One group blamed for a complex problem.",
    "Fear-mongering": "Predicts harm or catastrophe to move the reader.",
    "Dog Whistle": "Coded phrasing that signals something extra to an in-group.",
    "Us-vs-Them Framing": "Splits people into a virtuous us and a hostile them.",
    "Dehumanization": "People described as vermin, parasites, filth or cargo.",
    "Emotional Bait": "Asks for a reaction, share or outrage rather than making a point.",
    "Manufactured Urgency": "A deadline or now-or-never framing that is not real.",
    "Cherry Picking": "One favourable number or case stands in for the whole picture.",
    "Whataboutism": "Deflects by pointing at someone else's conduct.",
    "Gish Gallop": "Many separate assertions at once, too many to answer.",
    "Astroturfing": "Presents an organised campaign as spontaneous public feeling.",
    "False Dilemma": "Only two options offered when more exist.",
    "False Solution": "A complex problem is given a total, guaranteed fix.",
    "Ad Hominem": "Attacks the person instead of the argument.",
    "Hasty Generalization": "A sweeping rule drawn from one or two cases.",
    "Slippery Slope": "One step is said to lead inevitably to an extreme outcome.",
    "Straw Man": "Argues against a distorted version of the other side's position.",
    "Middle Ground": "Treats the midpoint between two claims as automatically correct.",
    "Bandwagon": "Everyone thinks this, therefore it is true.",
    "Appeal to Authority": "Cites status or a title in place of evidence.",
    "Loaded Question": "A question whose phrasing assumes the disputed fact.",
    "No True Scotsman": "Redefines the group to exclude an inconvenient example.",
    "Post Hoc": "Treats sequence as proof of cause.",
    "Motte and Bailey": "Advances a bold claim, retreats to a modest one when challenged.",
  };

  window.UF_TAXONOMY = {
    verdict(name) {
      return VERDICTS[String(name || "")] || FALLBACK;
    },

    // Labels the model invented rather than picked from the list arrive prefixed "Other: ".
    isUncategorised(name) {
      return String(name || "").startsWith("Other: ");
    },

    isHighRisk(name) {
      return HIGH_RISK.has(String(name || ""));
    },

    // One plain line explaining the label, or "" for an invented one we have no definition for.
    meaning(name) {
      return MEANINGS[String(name || "")] || "";
    },
  };
})();
